include/aluminum/traits/traits_base.hpp

```diff
@@ -0,0 +1,211 @@
+template <> struct OpSupportsAlgos<AlOperation::allgather> : std::true_type {};
+template <> struct OpSupportsAlgos<AlOperation::allgatherv> : std::true_type {};
+template <> struct OpSupportsAlgos<AlOperation::allreduce> : std::true_type {};
+template <> struct OpSupportsAlgos<AlOperation::alltoall> : std::true_type {};
+template <> struct OpSupportsAlgos<AlOperation::alltoallv> : std::true_type {};
+template <> struct OpSupportsAlgos<AlOperation::bcast> : std::true_type {};
+template <> struct OpSupportsAlgos<AlOperation::barrier> : std::true_type {};
+template <> struct OpSupportsAlgos<AlOperation::gather> : std::true_type {};
+template <> struct OpSupportsAlgos<AlOperation::gatherv> : std::true_type {};
+template <> struct OpSupportsAlgos<AlOperation::reduce> : std::true_type {};
+template <> struct OpSupportsAlgos<AlOperation::reduce_scatter> : std::true_type {};
+template <> struct OpSupportsAlgos<AlOperation::reduce_scatterv> : std::true_type {};
+template <> struct OpSupportsAlgos<AlOperation::scatter> : std::true_type {};
+template <> struct OpSupportsAlgos<AlOperation::scatterv> : std::true_type {};
+
+/** Identify the algorithm type for backend's operator. */
+template <AlOperation Op, typename Backend> struct OpAlgoType {};
+
+/** Helper for defining simple algorithm type structs. */
+#define AL_ADD_OP_ALGO_TYPE(op, backend)                    \
+  template <> struct OpAlgoType<AlOperation::op, backend> { \
+    using type = backend::op##_algo_type;                   \
+  }
+
+/**
+ * Define types supported by MPI.
+ *
+ * Note these are separate from the types the Aluminum MPI backend
+ * supports and refer to those supported specifically by MPI itself.
+ */
+template <typename T> struct IsTypeSupportedByMPI : std::false_type {};
+template <> struct IsTypeSupportedByMPI<char> : std::true_type {};
+template <> struct IsTypeSupportedByMPI<signed char> : std::true_type {};
+template <> struct IsTypeSupportedByMPI<unsigned char> : std::true_type {};
+template <> struct IsTypeSupportedByMPI<short> : std::true_type {};
+template <> struct IsTypeSupportedByMPI<unsigned short> : std::true_type {};
+template <> struct IsTypeSupportedByMPI<int> : std::true_type {};
+template <> struct IsTypeSupportedByMPI<unsigned int> : std::true_type {};
+template <> struct IsTypeSupportedByMPI<long> : std::true_type {};
+template <> struct IsTypeSupportedByMPI<unsigned long> : std::true_type {};
+template <> struct IsTypeSupportedByMPI<long long> : std::true_type {};
+template <> struct IsTypeSupportedByMPI<unsigned long long> : std::true_type {};
+template <> struct IsTypeSupportedByMPI<float> : std::true_type {};
+template <> struct IsTypeSupportedByMPI<double> : std::true_type {};
+template <> struct IsTypeSupportedByMPI<long double> : std::true_type {};
+#ifdef AL_HAS_HALF
+template <> struct IsTypeSupportedByMPI<__half> : std::true_type {};
+#endif
+#ifdef AL_HAS_BFLOAT
+template <> struct IsTypeSupportedByMPI<al_bfloat16> : std::true_type {};
+#endif
```

Comment:
Is this true? There's no `MPI_Datatype` value for these types and one cannot, say, `MPI_Reduce` with them without a custom reduction function, so it seems to me that either these should be `std::false_type` or the documentation should be clarified as to what "supported specifically by MPI itself" means.

Reply:
Good point. I'll clarify the documentation that these are supported via Aluminum magic. (Specifically: The `TypeMap` that everything uses will map them to a dummy datatype, and a custom reduction operator will be used.)